xzre.h

```diff
@@ -835,21 +842,23 @@ assert_offset(elf_handles_t, libc, 0x10);
 assert_offset(elf_handles_t, liblzma, 0x18);
 assert_offset(elf_handles_t, libcrypto, 0x20);
 
-typedef struct __attribute__((packed)) {
-	elf_handles_t *handles;
-	Elf64_Ehdr *libc_ehdr;
+typedef struct __attribute__((packed)) main_elf {
+	elf_handles_t *elf_handles;
+	Elf64_Ehdr *dynamic_linker_ehdr;
```

Comment:
I was a bit confused about this as this value was used to find the ld.so internal structs but then I realised __libc_stack_end is actually exported by ld.so (and not libc.so). The __tls_get_addr symbol that is used to find the initial ELF header is alsoa ld.so symbol

Reply:
Yeah, they are using `__tls_get_addr` to find roughly where the code section of `ld.so` is, so they can backtrack and find its ELF header.
Once they have that, they can get `r_debug` and look for loaded libraries (and at that point, they have full knowledge)